File: remotion-service/src/lib/analysis/reconcileTriggers.ts
import { type ChunkPlan, overlapZone } from "./planChunks";

export interface TimestampedTrigger {
  start: number;
  end?: number;
  confidence: number;
  [key: string]: unknown;
}

const DEDUPE_TOLERANCE_SECONDS = 1.5;

function triggerIdentity(item: TimestampedTrigger): string {
  const label = item.type ?? item.label ?? item.value ?? item.text ?? "";
  return String(label).toLowerCase();
}

function inZone(item: TimestampedTrigger, start: number, end: number): boolean {
  const endTs = item.end ?? item.start;
  const mid = (item.start + endTs) / 2;
  return mid >= start && mid <= end;
}

function isDuplicate(a: TimestampedTrigger, b: TimestampedTrigger): boolean {
  if (triggerIdentity(a) !== triggerIdentity(b)) {
    return false;
  }
  return Math.abs(a.start - b.start) <= DEDUPE_TOLERANCE_SECONDS;
}

/**
 * Deduplicate triggers detected in overlapping chunk windows.
 * Keeps highest-confidence entry when duplicates appear in overlap regions.
 */
export function reconcileTriggers<T extends TimestampedTrigger>(
  chunkOutputs: Array<{ chunk: ChunkPlan; triggers: T[] }>,
): T[] {
  if (chunkOutputs.length === 0) {
    return [];
  }
  if (chunkOutputs.length === 1) {
    return [...chunkOutputs[0].triggers];
  }

  const merged = chunkOutputs.flatMap(({ triggers }) => triggers);
  const overlapRanges: Array<{ start: number; end: number }> = [];
  for (let i = 0; i < chunkOutputs.length - 1; i += 1) {
    const zone = overlapZone(chunkOutputs[i].chunk, chunkOutputs[i + 1].chunk);
    if (zone) {
      overlapRanges.push(zone);
    }
  }

  const kept: T[] = [];

  for (const item of merged.sort((a, b) => a.start - b.start)) {
    const inOverlap = overlapRanges.some((z) => inZone(item, z.start, z.end));
    if (!inOverlap) {
      kept.push(item);
      continue;
    }

    const duplicateIdx = kept.findIndex((existing) => isDuplicate(existing, item));
    if (duplicateIdx >= 0) {
      if (item.confidence > kept[duplicateIdx].confidence) {
        kept[duplicateIdx] = item;
      }
      continue;
    }

    kept.push(item);
  }

  return kept.sort((a, b) => a.start - b.start);
}

File: apps/api/services/director/analysis/reconcile_triggers.py
"""Trigger reconciliation for chunked analysis (Phase 12)."""
from __future__ import annotations

from typing import Any

from services.director.analysis.plan_chunks import ChunkPlan, overlap_zone

_DEDUPE_TOLERANCE_SECONDS = 1.5


def _trigger_identity(item: dict[str, Any]) -> str:
    label = item.get("type") or item.get("label") or item.get("value") or item.get("text") or ""
    return str(label).lower()


def _in_zone(item: dict[str, Any], zone_start: float, zone_end: float) -> bool:
    start = float(item.get("start", 0))
    end = float(item.get("end", start))
    mid = (start + end) / 2
    return zone_start <= mid <= zone_end


def _is_duplicate(a: dict[str, Any], b: dict[str, Any]) -> bool:
    if _trigger_identity(a) != _trigger_identity(b):
        return False
    return abs(float(a.get("start", 0)) - float(b.get("start", 0))) <= _DEDUPE_TOLERANCE_SECONDS


def reconcile_triggers(
    chunk_outputs: list[tuple[ChunkPlan, list[dict[str, Any]]]],
) -> list[dict[str, Any]]:
    """Deduplicate triggers detected in overlapping chunk windows."""
    if not chunk_outputs:
        return []
    if len(chunk_outputs) == 1:
        return list(chunk_outputs[0][1])

    merged = [t for _, triggers in chunk_outputs for t in triggers]
    overlap_ranges: list[tuple[float, float]] = []
    for i in range(len(chunk_outputs) - 1):
        zone = overlap_zone(chunk_outputs[i][0], chunk_outputs[i + 1][0])
        if zone:
            overlap_ranges.append(zone)

    kept: list[dict[str, Any]] = []

    for item in sorted(merged, key=lambda t: float(t.get("start", 0))):
        in_overlap = any(_in_zone(item, z[0], z[1]) for z in overlap_ranges)
        if not in_overlap:
            kept.append(item)
            continue

        duplicate_idx = next(
            (idx for idx, existing in enumerate(kept) if _is_duplicate(existing, item)),
            -1,
        )
        if duplicate_idx >= 0:
            if float(item.get("confidence", 0)) > float(kept[duplicate_idx].get("confidence", 0)):
                kept[duplicate_idx] = item
            continue

        kept.append(item)

    return sorted(kept, key=lambda t: float(t.get("start", 0)))
